source/extensions/filters/network/thrift_proxy/filters/filter.h

```diff
@@ -154,6 +184,101 @@ class DecoderFilter : public virtual DecoderEventHandler {
 
 using DecoderFilterSharedPtr = std::shared_ptr<DecoderFilter>;
 
+/**
+ * Encoder filter interface.
+ *
+ * Currently the EncoderFilter and BidirectionFilter support
+ * a. peek the metadata and content for encoding,
+ * b. modify the metadata and content except string value for encoding, and
+ * c. what DecoderFilter supports for BidirectionFilter.
+ *
+ * Do not support
+ * a. pass through data separately for encode and decode, e.g., pass through data for request but
+ * not pass through data for response, and
+ * b. return StopIteration for decoder_events in for encoder filter and encode* events for
+ * bidirection filter. The filters trying to return StopIteration will reset the connection.
+ */
+class EncoderFilter : public FilterBase, public virtual DecoderEventHandler {
+public:
+  ~EncoderFilter() override = default;
+
+  /**
+   * Called by the connection manager once to initialize the filter encoder callbacks that the
+   * filter should use. Callbacks will not be invoked by the filter after onDestroy() is called.
+   */
+  virtual void setEncoderFilterCallbacks(EncoderFilterCallbacks& callbacks) PURE;
+
+  /**
+   * @return True if payload passthrough is supported. Called by the connection manager once after
+   * messageBegin.
+   */
+  virtual bool passthroughSupported() const PURE;
+};
+
+using EncoderFilterSharedPtr = std::shared_ptr<EncoderFilter>;
+
+/**
+ * Bidirection filter interface. @see EncoderFilter for limitation.
+ */
+class BidirectionFilter : public FilterBase {
```

Comment:
We should probably use "bidirectional" instead of bidirection, though I recognize that's a painful change.

Reply:
not that painful given current powerful IDE. I was hesitating to do the name change but I am happy that we have a second voice to change. Thanks!